api/src/main/java/jakarta/config/Config.java

```diff
@@ -30,27 +28,73 @@
+    String getKey();
+
+    /**
+     * Name of this node - the last element of a fully qualified key.
      * <p>
-     * The returned sources will be sorted by priority and name, which can be iterated in a thread-safe
-     * manner. The {@link java.lang.Iterable Iterable} contains a fixed number of {@link jakarta.config.spi.ConfigSource
-     * configuration
-     * sources}, determined at application start time, and the config sources themselves may be static or dynamic.
+     * For example for key {@code server.port} this method would return {@code port}.
+     *
+     * @return name of this node
+     */
+    String getName();
+
+    /**
+     * Single sub-node for the specified sub-key.
+     * For example if requested for key {@code server}, this method would return a config
+     * representing the {@code server} node, which would have for example a child {@code port}.
+     * The sub-key can return more than one level of nesting (e.g. using {@code server.tls} would
+     * return a node that contains the TLS configuration under {@code server} node).
+     *
+     * @param key sub-key to retrieve nested node.
+     * @return sub node, never null
+     */
+    Config get(String key);
+
+    /**
+     * Typed value created using a converter function.
+     * The converter is called only if this config node exists.
+     *
+     * @param converter to create an instance from config node
+     * @param <T> type of the object
+     * @return converted value of this node, or an empty optional if this node does not exist
+     * @throws java.lang.IllegalArgumentException if this config node cannot be converted to the desired type
+     */
+    <T> Optional<T> as(Function<Config, T> converter);
+
+    /**
+     * Typed value created using a discovered/built-in converter.
      *
-     * @return the configuration sources
+     * @param type class to convert to
+     * @param <T> type of the object
+     * @return converted value of this node, or an empty optional if this node does not exist
+     * @throws java.lang.IllegalArgumentException if this config node cannot be converted to the desired type
      */
-    Iterable<ConfigSource> getConfigSources();
+    <T> Optional<T> as(Class<T> type);
 
     /**
-     * Return the {@link jakarta.config.spi.Converter} used by this instance to produce instances of the specified type from
-     * string values.
+     * Direct value of this node used for string converters.
      *
-     * @param <T>
-     *            the conversion type
-     * @param forType
-     *            the type to be produced by the converter
-     * @return an {@link java.util.Optional} containing the converter, or empty if no converter is available for the specified
-     *         type
-     */
-    <T> Optional<Converter<T>> getConverter(Class<T> forType);
+     * @return value of this node
+     */
+    Optional<String> asString();
+
+    /*
+     * Shortcut helper methods
+     */
+    default Optional<Integer> asInt() {
```

Comment:
IMO we should never support `Optional<Integer|Long|Float|Double|etc.>` as long as we have `Optional[Int|Long|Double]` which more concisely expresses this information.

Perhaps we should add convenience methods in a follow-up PR?

Reply:
Agreed with @dmlloyd here ! I don't think `Int` type is in the top 2 types. Once we add this, it might open the gate for a potential bloated interface. Besides, it is so easy to get an `int` anyway.

Reply:
@Emily-Jiang  what would you think are the "top 2 types"?

Reply:
I have removed the shortcut method, we can come back to this later.

Reply:
@tomas-langer I think the "top 2 types" depends on the nature of the configurations. If the config sources are money related accounts, the top 2 types will be string and float. Therefore, I think removing the guessed top 2 types. Anyway, I think we are all good.